Skipping songs can cause a panic

New player logic in master branch can hit a panic when skipping passed the end of the play queue.

Steps to reproduce:

1. Start jamsonic with the `-experimental` flag.
2. Start playing the second to last song in an album.
3. Press "b" twice. Instead of just stopping the player tries to play an entry out of range which causes a panic.

Crash output:
```
panic: runtime error: makeslice: len out of range

goroutine 4 [running]:
github.com/TcM1911/jamsonic.(*playqueue).popSong(0xc42022e000, 0x0)
        /[REDACTED]/TcM1911/jamsonic/player.go:331 +0xa5
github.com/TcM1911/jamsonic.(*playqueue).(github.com/TcM1911/jamsonic.popSong)-fm(0xc4201c7c48)
        /[REDACTED]/TcM1911/jamsonic/player.go:182 +0x2a
github.com/TcM1911/jamsonic.(*Player).playNextInQueue(0xc4201c7c00, 0xc4207e1df0)
        /[REDACTED]/TcM1911/jamsonic/player.go:260 +0x4d
github.com/TcM1911/jamsonic.(*Player).playerLoop(0xc4201c7c00)
        /[REDACTED]/TcM1911/jamsonic/player.go:211 +0x78f
created by github.com/TcM1911/jamsonic.NewPlayer
        /[REDACTED]/TcM1911/jamsonic/player.go:63 +0x2f4
```

--- a/config.go
+++ b/config.go
@@ -20,5 +20,15 @@
 
 package jamsonic
 
+import (
+	"time"
+)
+
 var Experimental = false
 var Debug = false
+
+// BufferingWait is the time in milliseconds to wait on reading from the network socket
+// before playing the track. If this is to low, an EOF can be returned when reading
+// from the memory buffer causing the track from being skipped.
+// Default value is 200 ms.
+var BufferingWait = time.Duration(200 * time.Millisecond)

--- a/player.go
+++ b/player.go
@@ -22,6 +22,7 @@ package jamsonic
 
 import (
 	"bytes"
+	"errors"
 	"io"
 	"sync"
 	"time"
@@ -37,6 +38,12 @@ const (
 	Playing
 	// Paused is the state when the player has been paused.
 	Paused
+)
+
+var (
+	// ErrNoNextTrack is returned when the playing queue does not have a track to play
+	// but is asked to play one.
+	ErrNoNextTrack = errors.New("no track in playing queue")
 )
 
 // NewPlayer returns a new Player. The Provider should be a music provider.
@@ -217,7 +224,12 @@ func (p *Player) playerLoop() {
 			if ct != nil {
 				p.played.pushSong(ct)
 			}
-			p.playNextInQueue(p.queue.popSong)
+			err := p.playNextInQueue(p.queue.popSong)
+			if err == ErrNoNextTrack {
+				// If no next track, keep playing the current.
+				p.Error <- err
+				continue
+			}
 			songStart = time.Now()
 			pausedDuration = time.Duration(0)
 		case <-p.prevChan:
@@ -264,16 +276,19 @@ func (p *Player) playerLoop() {
 	}
 }
 
-func (p *Player) playNextInQueue(getTrack func() *Track) {
+func (p *Player) playNextInQueue(getTrack func() *Track) error {
 	p.queueMu.Lock()
 	ct := getTrack()
-	p.updateCurrentTrack(ct)
 	p.queueMu.Unlock()
+	if ct == nil {
+		return ErrNoNextTrack
+	}
+	p.updateCurrentTrack(ct)
 
 	stream, err := p.provider.GetStream(ct.ID)
 	if err != nil {
 		handleStreamError(p, err)
-		return
+		return nil
 	}
 	// Ensure we have control of this pointer.
 	p.bufMu.Lock()
@@ -315,6 +330,7 @@ func (p *Player) playNextInQueue(getTrack func() *Track) {
 	if err != nil {
 		handleStreamError(p, err)
 	}
+	return nil
 }
 
 func (p *Player) stopPlaying() {
@@ -360,6 +376,10 @@ func (q *playqueue) nextSong() *Track {
 // popSong returns the next song in the play queue and removes it from the queue.
 func (q *playqueue) popSong() *Track {
 	track := q.nextSong()
+	// If no next song, return nil
+	if track == nil {
+		return nil
+	}
 	q.arrayMu.Lock()
 	defer q.arrayMu.Unlock()
 	tmp := make([]*Track, len(q.array)-1)

--- a/player.go
+++ b/player.go
@@ -292,41 +292,37 @@ func (p *Player) playNextInQueue(getTrack func() *Track) error {
 	}
 	// Ensure we have control of this pointer.
 	p.bufMu.Lock()
-	tmp := p.buffer
 	// If buffered, just reset and reuse the buffer.
-	tmp.bufferedMu.Lock()
-	if tmp.buffered {
-		tmp.Reset()
+	p.buffer.bufferedMu.Lock()
+	if p.buffer.buffered {
+		p.buffer.Reset()
+		p.buffer.bufferedMu.Unlock()
 	} else {
 		// Since we can't stop to copy, create a new buffer
 		// and let the GC clean up the old buffer.
+		p.buffer.bufferedMu.Unlock()
 		p.buffer = newBufReadWriter()
 	}
-	tmp.bufferedMu.Unlock()
+	buf := p.buffer
 	p.bufMu.Unlock()
 
 	// Get the track data off the wire and into a memory buffer.
 	go func() {
-		p.bufMu.Lock()
-		writer := p.buffer
-		p.bufMu.Unlock()
-		_, cpErr := io.Copy(writer, stream)
+		_, cpErr := io.Copy(buf, stream)
 		if cpErr != nil {
 			p.Error <- cpErr
 			return
 		}
-		writer.bufferedMu.Lock()
-		writer.buffered = true
-		writer.bufferedMu.Unlock()
+		buf.bufferedMu.Lock()
+		buf.buffered = true
+		buf.bufferedMu.Unlock()
 		closeErr := stream.Close()
 		if closeErr != nil {
 			p.Error <- closeErr
 		}
 	}()
-	p.bufMu.Lock()
-	reader := p.buffer
-	p.bufMu.Unlock()
-	err = p.handler.Play(reader)
+	time.Sleep(BufferingWait)
+	err = p.handler.Play(buf)
 	if err != nil {
 		handleStreamError(p, err)
 	}

--- a/player_test.go
+++ b/player_test.go
@@ -46,6 +46,11 @@ var tracks = []*Track{
 	&Track{ID: "4"},
 }
 
+func init() {
+	// Set to 0 since it's not needed for testing.
+	BufferingWait = time.Duration(0)
+}
+
 func TestPlayControl(t *testing.T) {
 	assert := assert.New(t)
 	player, _, _, _ := getPlayer()

--- a/player_test.go
+++ b/player_test.go
@@ -626,6 +626,48 @@ func TestStreamBuffering(t *testing.T) {
 
 		p.Close()
 	})
+
+	t.Run("Handle errors when closing stream", func(t *testing.T) {
+		finishedChan := make(chan struct{})
+		expectedError := errors.New("expected error")
+		provider := &mockProvider{
+			doGetStream: func(id string) (io.ReadCloser, error) {
+				return &recorder{streamID: id, closeErr: expectedError}, nil
+			},
+		}
+		handler := &mockStreaHandler{
+			doFinished: func() <-chan struct{} { return finishedChan },
+			doPlay:     func(io.Reader) error { return nil },
+			doStop:     func() {},
+			doPause:    func() {},
+			doContinue: func() {},
+		}
+		p := NewPlayer(provider, handler, nil, 0)
+		p.CreatePlayQueue(tracks)
+
+		p.Play()
+		err := <-p.Error
+		assert.Equal(expectedError.Error(), err.Error(), "Returned wrong error")
+		assert.Equal(tracks[0], p.CurrentTrack(), "Wrong track set as next")
+
+		p.Close()
+	})
+}
+
+func TestQueue(t *testing.T) {
+	assert := assert.New(t)
+
+	t.Run("no panic if skipping past end return error", func(t *testing.T) {
+		p, _, _, _ := getPlayer()
+		p.CreatePlayQueue(tracks)
+		p.Play()
+		p.Next()
+		p.Next()
+		p.Next()
+		p.Next()
+		err := <-p.Error
+		assert.Equal(ErrNoNextTrack, err, "Wrong error returned")
+	})
 }
 
 func getPlayer() (*Player, chan struct{}, *mockProvider, *mockStreaHandler) {
@@ -652,6 +694,7 @@ type recorder struct {
 	read2    bool
 	read3    bool
 	read4    bool
+	closeErr error
 }
 
 func (r *recorder) Read(b []byte) (int, error) {
@@ -693,6 +736,9 @@ func (r *recorder) Read(b []byte) (int, error) {
 }
 
 func (r *recorder) Close() error {
+	if r.closeErr != nil {
+		return r.closeErr
+	}
 	return nil
 }
 

